### PR TITLE
Localization sweep

### LINKED ISSUE #277: Sweep through UI strings to internationalize
This is a followup to #275. After we have that basic framework for using translated strings in the UI, we'll want to sweep through our existing code and convert it all to use that framework.

Initially this won't do much of anything from a user's perspective, because we won't have more than a handful of strings translated until we wire things up to a service where people can contribute such translations, #276.

Nevertheless it's useful to do this early, because it sets the pattern so that we naturally follow it in new code, rather than having more code that needs to change in an eventual sweep later. We may also come across places that need a `context` object and didn't before, or otherwise call for refactoring, and it's convenient to do that sooner so that other changes to the same code are based on the new structure from the beginning.

Other notes:
* Many of these spots are marked with `TODO(i18n)`, though many others are not.
* Some spots marked `TODO(i18n)` are for different needs like formatting dates and times. Those are out of scope of this issue.
* If some strings are especially difficult to wire up for translation for some reason, we can mark them with `TODO(i18n)` and also file followup issues, and then leave them out for this issue.

### PR DESCRIPTION
**Pull request by sirpengi**
I've gone through the code and found any hard-coded strings or other `TODO(i18n)` markers and wired them up to the localization framework.

For some more complicated translations that didn't have inherit access to a `BuildContext` (exception messages, or certain widget labels and tooltips) I opted for a pattern of attaching methods that take the `ZulipLocalizations` object so they could perform the translation when needed.

### CODE REVIEW

**Comment by gnprice** [comment]
Thanks! I was preoccupied today with the Sentry-induced crashes in zulip-mobile that we resolved with v27.213, so I didn't get to a detailed review of this.

A couple of high-level comments which will make a good first pass:
* Let's in general squash the change that adds a given string to the ARB file together with the change that starts consuming it in our code. That provides context when reading either part of the change, from seeing the complementary part of it there in the same commit.
* Many of these changes are nice and easy and boring, and it's fine to have those squashed together with each other. (Though I'd also be happy to have them split more finely.) But it looks like some are more substantive and interesting — in particular, the changes in `lib/api/`.

  It'd be good to have a separate commit for those `lib/api/` changes, and probably one or more others for the other groups of "more complicated translations" that have the pattern you mention.

**Comment by sirpengi** [comment]
@gnprice the PR has been split and bunched as requested, ready for review

File: lib/api/exception.dart [hunk 1]
```diff
@@ -11,13 +13,26 @@ sealed class ApiRequestException implements Exception {
   /// and its route name would be "getMessages".
   final String routeName;
 
-  /// A user-facing description of the error.
+  /// Server supplied user-facing description of the error.
+  ///
+  /// For [ZulipApiException] this may be supplied by the server as the `message`
+  /// property in the JSON response.
```
Comment (gnprice) [inline lib/api/exception.dart:19]: Why the change from `is` to `may be`? Seems like it still is.

File: lib/api/exception.dart [hunk 2]
```diff
@@ -11,13 +13,26 @@ sealed class ApiRequestException implements Exception {
...
+  // TODO(i18n): no method yet to translate a server provided string.
+  final String? message;
```
Comment (gnprice) [inline lib/api/exception.dart:21]: As the meaning of this is changing to be more specific, it should get a more specific name to match. `serverMessage` would fit this description well.

File: lib/api/exception.dart [hunk 3]
```diff
@@ -11,13 +13,26 @@ sealed class ApiRequestException implements Exception {
...
+  final String Function(ZulipLocalizations)? translatableMessage;
+
+  ApiRequestException({required this.routeName, required this.message, required this.translatableMessage});
 
-  ApiRequestException({required this.routeName, required this.message});
+  String toTranslatedString(ZulipLocalizations zulipLocalizations) {
+    return (translatableMessage != null)
+      ? translatableMessage!(zulipLocalizations)
+      : super.toString();
+  }
```
Comment (gnprice) [inline lib/api/exception.dart:35]: I kind of think that this should become the only public way to read the error message, so that the two fields that back it would become private. It seems like in this design, it'd be a bug if we ever take the plain `message` and show that to the user, right? And that'd be a very easy bug to have.

In fact, asking the IDE to show references to the field, it seems like in this version of the PR we already introduce that bug — there's a reference to `e.message` in `lib/widgets/login.dart`:
```dart
        final message = (e is ZulipApiException)
          ? zulipLocalizations.loginServerErrorMessage(e.toTranslatedString(zulipLocalizations))
          : e.message;
        showErrorDialog(context: context,
          title: zulipLocalizations.loginServerLoginFailed,
          message: message);
```

Then the name of this method can also become something simpler, like `message`.

File: lib/widgets/action_sheet.dart [hunk 4]
```diff
@@ -43,18 +44,19 @@ abstract class MessageActionSheetMenuItemButton extends StatelessWidget {
   }) : assert(messageListContext.findAncestorWidgetOfExactType<MessageListPage>() != null);
 
   IconData get icon;
-  String get label;
+  String Function(ZulipLocalizations) get translateLabel;
```
Comment (gnprice) [inline lib/widgets/action_sheet.dart:47]: I think the name `label` is still good for this.

The background principle is that translation/i18n should be the normal and expected situation, for any string that appears in the UI. So this is just the method that provides the label. It's now a function that takes an argument — so not just a getter returning `String` — but that's just how it has to be in a world where the UI is appropriately internationalized, so it doesn't call for being highlighted in the name.

Comment (gnprice) [inline lib/widgets/action_sheet.dart:47]: Separately, it doesn't seem like this benefits from being a getter, now that it's returning a function — it should just be a method, now taking an argument but still returning a `String`.

File: assets/l10n/app_en.arb [hunk 5]
```diff
@@ -27,18 +31,290 @@
...
+  "@actionSheetQuoteAndReply":  {
+    "description": "Label for Quote and reply button on action sheet."
+  },
+  "actionSheetQuotationFailed": "Quotation failed",
+  "@actionSheetQuotationFailed": {
+    "description": "Error dialog message when quoting a message failed."
+  },
+  "actionSheetShare": "Share",
+  "@actionSheetShare":  {
+    "description": "Label for share button on action sheet."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:84]: I like the "action sheet" prefix on these names. Perhaps a bit more explicitness to further scope what the context is about: `actionSheetOptionShare`.

File: assets/l10n/app_en.arb [hunk 6]
```diff
@@ -27,18 +31,290 @@
...
+  "@permissionsReadExternalStorageDenied": {
+    "description": "Message for dialog when the user needs to grant permissions for external storage read access."
+  },
+  "actionSheetCopyMessageText": "Copy message text",
+  "@actionSheetCopyMessageText":  {
+    "description": "Label for copy message text button on action sheet."
+  },
+  "actionSheetCouldNotFetchMessageSource": "Could not fetch message source",
+  "@actionSheetCouldNotFetchMessageSource": {
+    "description": "Dialog message when the source of a message could not be fetched."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:60]: For these error messages, I feel like the relevant context isn't so much "action sheet" as that it's the error message to appear in a dialog box.

As a test for making these naming choices, consider if we had some other place in the UI — somewhere other than an action sheet — where you could do something that needed to fetch the source of a message, and that fetching failed. Would we want the message to be different? I think we'd want it to be the same (and therefore save translators' effort re-translating it, and avoid showing users untranslated versions until they do).

So perhaps `errorCouldNotFetchMessageSource`. Potentially could be like `errorDialogCouldNotFetchMessageSource`, but I think the "dialog" specificity wouldn't be helpful either — if we show it in a toast/[snackbar](https://m3.material.io/components/snackbar/overview) or something, we'll still want the same message.

File: assets/l10n/app_en.arb [hunk 7]
```diff
@@ -27,18 +31,290 @@
...
+  "@contentValidationErrorQuoteAndReplyInProgress": {
+    "description": "Content validation error message when a quotation has not completed yet."
+  },
+  "contentValidationErrorUploadInProgress": "Please wait for the upload to complete.",
+  "@contentValidationErrorUploadInProgress": {
+    "description": "Content validation error message when attachments have not finished uploading."
+  },
+  "dialogButtonOK": "OK",
+  "@dialogButtonOK": {
+    "description": "Button label in generic dialogs to confirm."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:216]: The actual use of this is somewhat more specific — it's in an *error* dialog, as the only available action, to acknowledge the error and dismiss the dialog.

I think that's a relevant distinction: it's fairly likely that if we had a dialog box with an actual choice, with options like perhaps "OK" / "Cancel", then in some languages the "OK" would be translated differently from the "OK" of the error dialog.

File: assets/l10n/app_en.arb [hunk 8]
```diff
@@ -27,18 +31,290 @@
...
+  },
+  "dialogButtonCancel": "Cancel",
+  "@dialogButtonCancel": {
+    "description": "Button label in generic dialogs to cancel."
+  },
+  "dialogButtonContinue": "Continue",
+  "@dialogButtonContinue": {
+    "description": "Button label in generic dialogs to continue."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:224]: Similarly these names and descriptions should be a bit more specific.

The function name is `showSuggestedActionDialog`. So e.g. "to continue with a suggested action" / "to cancel a suggested action".

**Review by gnprice — COMMENTED** [review]
Thanks! That split is very helpful.

Comments below. I read the `lib/api/` commit fully, and skimmed through the rest. Several of the points below apply in several places across several commits; taken that way, I think they broadly cover all the feedback I'd have.

File: lib/api/exception.dart (hunk 1 above)
Comment (sirpengi) [reply lib/api/exception.dart:19]: You're right, the `may be` applied to the broader `ApiRequestConnection` class, where in some cases the `message` was supplied by the server, but that is the case specifically of `ZulipApiException`

File: lib/api/exception.dart [hunk 9]
```diff
@@ -15,9 +17,10 @@ sealed class ApiRequestException implements Exception {
   ///
   /// For [ZulipApiException] this is supplied by the server as the `message`
   /// property in the JSON response, and is translated into the user's language.
-  final String message;
+  final String Function(ZulipLocalizations) message;
```
Comment (gnprice) [inline lib/api/exception.dart:20]: nit in commit message:

    core: Add translations for exceptions related to api requests

A better prefix would be `api:`. In particular "core" doesn't really mean anything here unless it's scoped within "api" (like the filename `lib/api/core.dart`) — there's lots of pieces of the codebase that could be thought of as "core".

File: test/api/core_test.dart [hunk 10]
```diff
@@ -153,9 +154,13 @@ void main() {
...
-    checkRequest(http.ClientException('Oops'), it()..message.equals('Oops'));
-    checkRequest(const TlsException('Oops'), it()..message.equals('Oops'));
-    checkRequest((foo: 'bar'), it()..message.equals('Network request failed'));
+    final zulipLocalizations = lookupZulipLocalizations(ZulipLocalizations.supportedLocales.first);
+    checkRequest(http.ClientException('Oops'), it()..message(zulipLocalizations).equals('Oops'));
+    checkRequest(const TlsException('Oops'), it()..message(zulipLocalizations).equals('Oops'));
+    final expectedMessage = zulipLocalizations.apiConnectionNetworkRequestFailed;
+    checkRequest((foo: 'bar'), it()
+      ..message(zulipLocalizations)
+      .equals(expectedMessage));
```
Comment (gnprice) [inline test/api/core_test.dart:163]: This bit seems cleanest to read if `expectedMessage` is inlined. Shorter; more parallel to the other examples just above; and the last line gets longer, but it's under 80 columns so that in itself is no problem at all.

File: test/api/exception_checks.dart [hunk 11]
```diff
@@ -1,9 +1,11 @@
 import 'package:checks/checks.dart';
+import 'package:flutter_gen/gen_l10n/zulip_localizations.dart';
 import 'package:zulip/api/exception.dart';
 
 extension ApiRequestExceptionChecks on Subject<ApiRequestException> {
   Subject<String> get routeName => has((e) => e.routeName, 'routeName');
-  Subject<String> get message => has((e) => e.message, 'message');
+  Subject<String> message (ZulipLocalizations zulipLocalizations) =>
```
Comment (gnprice) [inline test/api/exception_checks.dart:7]: nit: no space between function name and open-paren

File: lib/widgets/action_sheet.dart [hunk 12]
```diff
@@ -43,18 +44,19 @@ abstract class MessageActionSheetMenuItemButton extends StatelessWidget {
   }) : assert(messageListContext.findAncestorWidgetOfExactType<MessageListPage>() != null);
 
   IconData get icon;
-  String get label;
+  String Function(ZulipLocalizations) get label;
```
Comment (gnprice) [inline lib/widgets/action_sheet.dart:47]: bump https://github.com/zulip/zulip-flutter/pull/306#discussion_r1336547833

File: assets/l10n/app_en.arb [hunk 13]
```diff
@@ -27,18 +31,298 @@
...
+  "@contentValidationErrorUploadInProgress": {
+    "description": "Content validation error message when attachments have not finished uploading."
+  },
+  "suggestedActionDialogCancel": "Cancel",
+  "@suggestedActionDialogCancel": {
+    "description": "Button label to cancel a suggested action."
+  },
+  "suggestedActionDialogContinue": "Continue",
+  "@suggestedActionDialogContinue": {
+    "description": "Button label in accept a suggested action."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:256]: seems like some words missing

File: assets/l10n/app_en.arb [hunk 14]
```diff
@@ -27,18 +31,298 @@
...
+  "@errorLoginFailed": {
+    "description": "Error dialog title when login for a Zulip server fails."
+  },
+  "errorMessageNotSent": "Message not sent",
+  "@errorMessageNotSent": {
+    "description": "Message for compose box when a message could not be sent."
+  },
+  "errorMessageDoesNotSeemToExist": "That message does not seem to exist.",
+  "@errorMessageDoesNotSeemToExist": {
+    "description": "Error message when loading a message that does not exist."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:126]: nit: This description gets introduced as one thing by an earlier commit, then edited by an otherwise-unrelated later commit.

File: assets/l10n/app_en.arb [hunk 15]
```diff
@@ -49,6 +333,42 @@
       }
     }
   },
+  "serverExceptionMalformedResponse": "Server gave malformed response; HTTP status {httpStatus}",
+  "@serverExceptionMalformedResponse": {
+    "description": "Message for error dialog when an API call fails because we could not parse it.",
+    "placeholders": {
+      "httpStatus": {
+        "type": "int",
+        "example": "500"
```
Comment (gnprice) [inline assets/l10n/app_en.arb:342]: nit: this is unrepresentative because for HTTP status 500 we'll never use this message (see the `MalformedServerResponseException` constructor call sites, and the `httpStatus` logic in `_makeApiException`). Probably most typical is 200:
```suggestion
  "serverExceptionMalformedResponse": "Server gave malformed response; HTTP status {httpStatus}",
  "@serverExceptionMalformedResponse": {
    "description": "Message for error dialog when an API call fails because we could not parse it.",
    "placeholders": {
      "httpStatus": {
        "type": "int",
        "example": "200"
```

File: assets/l10n/app_en.arb [hunk 16]
```diff
@@ -27,18 +31,298 @@
...
+  "@loginHidePassword": {
+    "description": "Icon label for button to hide password in input form."
+  },
+  "loginValidationPassword": "Please enter your password",
+  "@loginValidationPassword": {
+    "description": "Prompt for input for password field."
+  },
+  "loginValidationPasswordLabel": "Password",
+  "@loginValidationPasswordLabel": {
+    "description": "Label for input for password field."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:300]: Is this about validation? I think it isn't — it's just the label on the "password" field at login.

Similarly for email and username below.

File: assets/l10n/app_en.arb [hunk 17]
```diff
@@ -27,18 +31,298 @@
...
+  "@loginServerUrlInputLabel": {
+    "description": "Input label in login page for Zulip server URL entry."
+  },
+  "loginHidePassword": "Hide password",
+  "@loginHidePassword": {
+    "description": "Icon label for button to hide password in input form."
+  },
+  "loginValidationPassword": "Please enter your password",
+  "@loginValidationPassword": {
+    "description": "Prompt for input for password field."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:296]: I'd describe this as stronger than a prompt — it's an error message.

In particular it gets returned from a `validator` callback, which is documented (at [FormField.validator]) as:
```dart
  /// An optional method that validates an input. Returns an error string to
  /// display if the input is invalid, or null otherwise.
  ///
  /// The returned value is exposed by the [FormFieldState.errorText] property.
  /// The [TextFormField] uses this to override the [InputDecoration.errorText]
  /// value.
```
So the returned value is "an error string", exposed by an `errorText` property.

This specific error message is for when the user didn't enter a password in the password field. So could be `loginErrorMissingPassword`.

File: assets/l10n/app_en.arb [hunk 18]
```diff
@@ -27,18 +31,298 @@
...
+  "permissionsNeededOpenSettings": "Open settings",
+  "@permissionsNeededOpenSettings": {
+    "description": "Button label for permissions dialog button that opens the system settings screen."
+  },
+  "permissionsNeededCameraAccessDenied": "To upload an image, please grant Zulip additional permissions in Settings.",
+  "@permissionsNeededCameraAccessDenied": {
     "description": "Message for dialog when the user needs to grant permissions for camera access."
   },
+  "permissionsReadExternalStorageDenied": "To upload files, please grant Zulip additional permissions in Settings.",
+  "@permissionsReadExternalStorageDenied": {
```
Comment (gnprice) [inline assets/l10n/app_en.arb:51]: These should be consistent about whether "needed" is a part of their naming prefix.

File: assets/l10n/app_en.arb [hunk 19]
```diff
@@ -27,18 +31,298 @@
   "@profileButtonSendDirectMessage": {
     "description": "Label for button in profile screen to navigate to DMs with the shown user."
   },
-  "cameraAccessDeniedTitle": "Permissions needed",
-  "@cameraAccessDeniedTitle": {
-    "description": "Title for dialog when the user needs to grant permissions for camera access."
+  "permissionsNeededError": "Error",
+  "@permissionsNeededError": {
+    "description": "General title for access dialog when we encounter an unknown error."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:36]: This doesn't seem to get used.

File: assets/l10n/app_en.arb [hunk 20]
```diff
@@ -27,18 +31,298 @@
...
+        "type": "int",
+        "example": "4"
+      },
+      "maxFileUploadSizeMib": {
+        "type": "int",
+        "example": "15"
+      },
+      "listMessage": {
+        "type": "String",
+        "example": "foo.txt\nbar.txt"
```
Comment (gnprice) [inline assets/l10n/app_en.arb:98]: nit: make the example data consistent — if it's two filenames, then the number should be 2

File: assets/l10n/app_en.arb (hunk 18 above)
Comment (gnprice) [reply assets/l10n/app_en.arb:51]: The "needed" does feel kind of redundant with "denied". Perhaps `permissionDeniedFooBar`?

File: lib/widgets/compose_box.dart [hunk 21]
```diff
@@ -565,13 +583,14 @@ Future<Iterable<_File>> _getFilePickerFiles(BuildContext context, FileType type)
   });
 }
 
-class _AttachFileButton extends _AttachUploadsButton {
-  const _AttachFileButton({required super.contentController, required super.contentFocusNode});
+class _AttachFileButton extends _AttachUploadsButton { const _AttachFileButton({required super.contentController, required super.contentFocusNode});
```
Comment (gnprice) [inline lib/widgets/compose_box.dart:586]: This seems unintended :slightly_smiling_face:

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! Comments below.

Oh and a commit-message nit:
38498c471 msglist [nfc]: update i18n todo with specific issue
ea9c599b3 clipboard test [nfc]: Remove unnecessary TODO
3e20a176c core: Add translations for exceptions related to api requests
86753c1ec action_sheet: translations
5edc7a8b3 dialog: translations
990548f1b lightbox: translations
d49c1a6d4 login: translations
629e84abc compose: translations

Use sentence case in the summary lines, same as at https://github.com/zulip/zulip-flutter/pull/305#discussion_r1333513047 .

File: assets/l10n/app_en.arb (hunk 19 above)
Comment (sirpengi) [reply assets/l10n/app_en.arb:36]: There were some unreplaced strings this was meant to stand in for, so this is being used now.

**Comment by sirpengi** [comment]
@gnprice ready for another go!

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! This is definitely converging — I think at this point I'm happy with all the structure of the code here.

Comments below. Several of them are places where I spotted a discrepancy between the existing hard-coded string and the new English version. It should be doable to really have zero of those, by rereading through the draft changes; I'd like to get to that point before merge, because that's otherwise a potential way for typos and other errors to sneak into the UI.

Related to that is ways of keeping down the complexity of reviewing these changes. One thing that's contributed to making these changes as readable as they are is how you've generally kept the strings grouped thematically, and ordered within each group either to match the corresponding code or alphabetically; that's been very helpful.

I think the biggest other area that would help there is splitting up the biggest sets of changes into smaller commits. Smaller commits tend to be easier to review in general (so long as they're [coherent](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)); and the nature of these changes makes them especially sensitive to commit size, for reasons I explain in the last comment below.

Tha last comment is about one place to split: the biggest commit, which appears to correspond to a squash of the two biggest commits from the previous revision.

And then I think in fact it would help to split those two biggest commits up somewhat further. For example, in `compose_box` it could be validations; then the upload/attach flows; then other strings, including buttons and hint text.

File: assets/l10n/app_en.arb [hunk 22]
```diff
@@ -27,28 +27,317 @@
...
+  },
+  "errorMessageNotSent": "Message not sent",
+  "@errorMessageNotSent": {
+    "description": "Message for compose box when a message could not be sent."
+  },
+  "errorMessageDoesNotSeemToExist": "That message does not seem to exist.",
+  "@errorMessageDoesNotSeemToExist": {
+    "description": "Error message when loading a message that does not exist."
+  },
+  "errorServerErrorMessage": "The server said:\n\n{message}",
```
[the remark this answers is not in the document]
Comment (sirpengi) [reply assets/l10n/app_en.arb:109]: This was a duplicate (probably showed up when I moved things around) and it's strange the `gen_l10n` command didn't flag it (and it doesn't look like there's a config option for that either)

File: lib/widgets/compose_box.dart [hunk 23]
```diff
@@ -535,23 +544,25 @@ Future<Iterable<_File>> _getFilePickerFiles(BuildContext context, FileType type)
...
+        message: zulipLocalizations.permissionsDeniedReadExternalStorage,
+        actionButtonText: zulipLocalizations.permissionsNeededOpenSettings,
         onActionButtonPress: () {
           AppSettings.openAppSettings();
         });
     } else {
-      // TODO(i18n)
-      showErrorDialog(context: context, title: 'Error', message: e.toString());
+      showErrorDialog(context: context,
+        title: zulipLocalizations.permissionsNeededError,
```
[the remark this answers is not in the document]
Comment (sirpengi) [reply lib/widgets/compose_box.dart:564]: Opted to reuse an existing `errorDialogTitle` that already existed that fits use-case

File: lib/widgets/compose_box.dart [hunk 24]
```diff
@@ -48,12 +48,12 @@ enum TopicValidationError {
   mandatoryButEmpty,
   tooLong;
```
[the remark this answers is not in the document]
Comment (sirpengi) [reply lib/widgets/compose_box.dart:50]: This was two commits but got squashed (I believe unintentionally). I've broken it up into further conceptual pieces.

**Comment by sirpengi** [comment]
@gnprice Thanks for the comments as always. This is ready for review again!

File: assets/l10n/app_en.arb [hunk 25]
```diff
@@ -27,28 +27,298 @@
...
+  "@loginPageTitle": {
+    "description": "Page title for login page."
+  },
+  "loginFormSubmitLabel": "Log in",
+  "@loginFormSubmitLabel": {
+    "description": "Button text to submit login credentials."
+  },
+  "loginAddAnAccount": "Add an account",
+  "@loginAddAnAccount": {
+    "description": "Page title for screen to add a Zulip account."
```
Comment (gnprice) [inline assets/l10n/app_en.arb:239]: I like the convention you've used elsewhere of having the name end in "page title" when it's the title of a page. Let's apply that consistently here, too.

File: lib/widgets/action_sheet.dart [hunk 26]
```diff
@@ -104,13 +109,14 @@ Future<String?> fetchRawContentWithFeedback({
...
     //   On final failure or success, auto-dismiss the snackbar.
+    final zulipLocalizations = ZulipLocalizations.of(context);
     try {
       fetchedMessage = await getMessageCompat(PerAccountStoreWidget.of(context).connection,
         messageId: messageId,
         applyMarkdown: false,
       );
       if (fetchedMessage == null) {
-        errorMessage = 'That message does not seem to exist.';
+        errorMessage = zulipLocalizations.errorMessageDoesNotSeemToExist;
```
Comment (gnprice) [inline lib/widgets/action_sheet.dart:119]: nit: These three commit messages should have a verb:

8b9ea0fb8 action_sheet: Translations
159c85421 dialog: Translations
d4b0cd2b4 lightbox: Translations

"Add translations", like in several of your other commits, would do nicely.

**Review by gnprice — COMMENTED** [review]
Thanks @sirpengi for the revision! This is very close now to merge — just a few comments on these changes, all minor. I appreciate the additional splitting-up of the commits; that was quite helpful for reading them.

There are also some small wording tweaks I have for the descriptions, which I can just do as a commit on top at merge time.

The one other comment I have is that this PR shouldn't be marked as closing #277, because there are still some hard-coded English strings left — see in particular `git grep TODO.i18n` and `git grep 'Text\(['\''"]' lib/`. But let's leave those out of scope for this PR. Perhaps the right approach is that after we merge this, I'll take a turn going through the remaining strings; that will also be good for spreading around our experience with the workflow.

File: assets/l10n/app_en.arb (hunk 25 above)
Comment (gnprice) [reply assets/l10n/app_en.arb]: bump -- "page title" :-)

**Comment by gnprice** [comment]
Thanks for the revison! Merging, with that added commit on top tweaking some of the descriptions.

(And I guess the small last round of review that your last revision was in response to didn't make it into the GitHub thread's main timeline; but it's up at https://github.com/zulip/zulip-flutter/pull/306#discussion_r1346095334 .)

I'll also go add myself as an assignee on the issue #277, and will plan to pick up the remaining part of it soon.
